pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +237,306 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+		Commit:                lag.Commit,
+		CommitRecordTimestamp: time.UnixMilli(commitRecTs),
+		LastSeenOffset:        lastSeenOffset,
+		LastBlockEnd:          time.UnixMilli(lastBlockEndTs),
+	}
+}
+
+// consumePartition consumes records from the given partition until the cycleEnd timestamp.
+// If the partition is lagging behind, it takes care of calling consuming it in sections.
+func (b *BlockBuilder) consumePartition(ctx context.Context, partition int32, state partitionState, cycleEnd time.Time) (err error) {
+	builder := NewTSDBBuilder(b.logger, b.cfg.DataDir, b.cfg.BlocksStorage, b.limits, b.tsdbBuilderMetrics)
+	defer runutil.CloseWithErrCapture(&err, builder, "closing tsdb builder")
+
+	// Section is a portion of the partition to process in a single pass. One cycle may process multiple sections if the partition is lagging.
+	sectionEnd := cycleEnd
+	if sectionEnd.Sub(state.CommitRecordTimestamp) > time.Duration(1.5*float64(b.cfg.ConsumeInterval)) {
+		// We are lagging behind by more than 1.5*interval or there is no commit. We need to consume the partition in sections.
+		// We iterate through all the ConsumeInterval intervals, starting from the first one after the last commit until the cycleEnd,
+		// i.e. [T, T+interval), [T+interval, T+2*interval), ... [T+S*interval, cycleEnd)
+		// where T is the CommitRecordTimestamp, the timestamp of the record, whose offset we committed previously.
+		// When there is no kafka commit, we play safe and assume LastSeenOffset, and LastBlockEnd were 0 to not discard any samples unnecessarily.
+		sectionEnd, _ = nextCycleEnd(
+			state.CommitRecordTimestamp,
+			b.cfg.ConsumeInterval,
+			b.cfg.ConsumeIntervalBuffer,
+		)
+		level.Info(b.logger).Log("msg", "partition is lagging behind the cycle", "partition", partition, "lag", state.Lag, "section_cycle_end", sectionEnd, "cycle_end", cycleEnd, "commit_rec_ts", state.CommitRecordTimestamp)
+	}
+	for !sectionEnd.After(cycleEnd) {
+		partitionLogger := log.With(b.logger, "partition", partition, "lag", state.Lag, "section_cycle_end", sectionEnd)
+		state, err = b.consumePartitionSection(ctx, partitionLogger, builder, partition, state, sectionEnd)
+		if err != nil {
+			return fmt.Errorf("consume partition %d: %w", partition, err)
+		}
+		sectionEnd = sectionEnd.Add(b.cfg.ConsumeInterval)
+	}
+
+	return nil
+}
+
+func (b *BlockBuilder) consumePartitionSection(ctx context.Context, logger log.Logger, builder *TSDBBuilder, partition int32, state partitionState, sectionEnd time.Time) (_ partitionState, retErr error) {
+	// Oppose to the section's range (and cycle's range), that include the ConsumeIntervalBuffer, the block's range doesn't.
+	// Thus, truncate the timestamp with ConsumptionInterval here to round the block's range.
+	blockEnd := sectionEnd.Truncate(b.cfg.ConsumeInterval)
+
+	var numBlocks int
+	defer func(t time.Time, startState partitionState) {
+		dur := time.Since(t)
+
+		// Don't propagate any errors derived from the cancelled context.
```

Comment:
I'm not convinced it's correct to not propagate the `context.Canceled` error, for two reasons:

1. Design: we're used to propagate it to the caller, so they can take an informed decision
2. The caller doesn't stop the loop with the current code